src/solver/conv_hip_implicit_gemm_grouped_fwd_xdlops.cpp

```diff
@@ -189,23 +193,143 @@ bool ConvHipImplicitGemmGroupFwdXdlops::CheckCKApplicability(
 
+std::vector<std::string> GetKernelAsTokens(const std::string& kernel)
+{
+    std::vector<std::string> tokens;
+    std::string token;
+    std::istringstream tokenStream(
+        kernel.substr(kernel.find('<') + 1, kernel.find('>') - kernel.find('<') - 1));
+    while(std::getline(tokenStream, token, ','))
+    {
+        token.erase(remove_if(token.begin(), token.end(), isspace),
+                    token.end()); // strip whitespace
+        tokens.push_back(token);
+    }
+    return tokens;
+}
+
+void PerformanceConfigHipImplicitGemmGroupFwdXdlops::InitHeuristicKernelIDs()
+{
+    for(int i = 0; i < valid_kernels.size(); i++)
+    {
+        if(valid_kernels[i].find("DeviceGroupedConvFwdMultipleD_Xdl_CShuffle") != std::string::npos)
+        {
+            heuristic_indexes.push_back(i);
+            heuristic_kernels.push_back(GetKernelAsTokens(valid_kernels[i]));
+        }
+    }
+}
+
+bool PerformanceConfigHipImplicitGemmGroupFwdXdlops::ModelApplyToken(int idx, std::string value)
+{
+
+    std::vector<int> new_heuristic_indexes;
+    new_heuristic_indexes.reserve(heuristic_indexes.size());
+    if(idx >= 5)
+        idx += 2; // skip MPerXDL and NPerXDL as they are constant
+    std::copy_if(
+        heuristic_indexes.begin(),
+        heuristic_indexes.end(),
+        std::back_inserter(new_heuristic_indexes),
+        [&](int heuristic_index) { return heuristic_kernels[heuristic_index][idx] == value; });
+    if(new_heuristic_indexes.empty())
+        return false;
+    heuristic_indexes = new_heuristic_indexes;
+    return true;
+}
+
+static std::vector<float> GetFeatures(const ProblemDescription& problem, std::size_t num_cu)
+{
+    std::size_t n = 18;
+    std::vector<float> features(n, 0.0f);
+    features[0]  = problem.GetInDataType() == miopenFloat ? 2 : 1;
+    features[1]  = problem.GetInChannels_();
+    features[2]  = problem.GetInHeight_();
+    features[3]  = problem.GetInWidth_();
+    features[4]  = problem.GetOutChannels_();
+    features[5]  = problem.GetOutHeight_();
+    features[6]  = problem.GetOutWidth_();
+    features[7]  = problem.GetWeightsHeight_();
+    features[8]  = problem.GetWeightsWidth_();
+    features[9]  = problem.GetPadH();
+    features[10] = problem.GetPadW();
+    features[11] = problem.GetKernelStrideH();
+    features[12] = problem.GetKernelStrideW();
+    features[13] = problem.GetDilationH();
+    features[14] = problem.GetDilationW();
+    features[15] = problem.GetBatchSize_();
+    features[16] = problem.GetGroupCount();
+    features[17] = num_cu;
+    return features;
+}
+
+template <typename DataType>
+void PerformanceConfigHipImplicitGemmGroupFwdXdlops::RunParameterPredictionModel(
+    const ExecutionContext& ctx, const ProblemDescription& problem)
+{
+    valid_kernels = FillValidKernelsIDs<DeviceOpGFwdPtrs<DataType>, CKArgs>(
+        problem); // filter valid_kernel ID's
+    InitHeuristicKernelIDs();
+    static const std::string& arch  = ctx.GetStream().GetDeviceName();
```

Comment:
Is it for caching purposes?

Reply:
The static const is because the name of the architecture will not change during a run of MIOpen, so calling GetStream().GetDeviceName() is redundant as the architecture should remain the same across all instances of PerformanceConfigHipImplicitGemmGroupFwdXdlops.

Reply:
To follow up on this, `ai::tuning::GetModel` checks for whether the architecture has changed. 

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/ffbc4280e2b309574e41241ea8527cd1277c8e48/src/conv/heuristics/ai_heuristics.cpp#L495-L513

But since `arch` is declared `static`, that check is broken, right?

Reply:
@msaudulhassan This was a check suggested by @DrizztDoUrden. prevArch should only be initialized the first time and if there are subsequent calls to GetModel for architectures that do not match the initial architecture that was passed in we would throw an error. However I didn't understand exactly why this was necessary as I thought this would be impossible since a run of MIOpen can only run on a single gpu instance/architecture.

Reply:
Yes, I too am unclear about as to when such a case would come into play, but I wonder if we can handle it instead of throwing an error. What if we make `models` a 2D dictionary, indexed first by `arch` and then by `solver`?

Reply:
Another point: in this function
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/ffbc4280e2b309574e41241ea8527cd1277c8e48/src/solver/conv_hip_implicit_gemm_grouped_fwd_xdlops.cpp#L266-L289
on line 
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/ffbc4280e2b309574e41241ea8527cd1277c8e48/src/solver/conv_hip_implicit_gemm_grouped_fwd_xdlops.cpp#L273
you are defining arch as `static`, and then on line
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/ffbc4280e2b309574e41241ea8527cd1277c8e48/src/solver/conv_hip_implicit_gemm_grouped_fwd_xdlops.cpp#L276
`arch` is passed to `ModelSetParams()`. On line
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/ffbc4280e2b309574e41241ea8527cd1277c8e48/src/conv/heuristics/ai_heuristics.cpp#L521
`ModelSetParams()` passes `arch` to `GetModel()`.  So, `GetModel()` is getting a cached arch, which means `prevArch` would _always_ equal `arch`

Reply:
@msaudulhassan I will remove this check in the GetModel() function as it seems redundant.